rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        let acquire_result = self
+            .internal_acquire(lease_expiration, request_timeout)
+            .await;
+
+        if renewal_period.is_some() {
+            self.auto_renewal_cancellation_token = CancellationToken::new();
+            let mut self_clone = self.clone();
+
+            tokio::task::spawn({
+                async move {
+                    loop {
+                        select! {
+                            () = self_clone.auto_renewal_cancellation_token.cancelled() => {
+                                break; // Auto-renewal is cancelled.
+                            }
+                            () = tokio::time::sleep(renewal_period.unwrap()) => {}
+                        }
+
+                        if self_clone
+                            .internal_acquire(lease_expiration, request_timeout)
+                            .await
+                            .is_err()
+                        {
+                            // Acquire failed. Stopping Auto-renewal.
+                            break;
+                        }
+                    }
+                }
+            });
+        }
+
+        acquire_result
+    }
+
+    /// Releases a lease if and only if requested by the lease holder (same client id).
+    ///
+    /// Note: `request_timeout` is rounded up to the nearest second.
+    ///
+    /// Returns `Ok()` if lease is no longer held by this `lease_holder`, or `Error` otherwise.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `V Delete` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// # Panics
+    /// Possible panic if `std::Lock::lock()` fails.
+    pub async fn release(&mut self, request_timeout: Duration) -> Result<(), Error> {
+        // Stop auto-renewal.
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        *self
+            .current_fencing_token
+            .lock()
+            .expect("Could not lock mutex") = None;
+
+        match self
+            .state_store
+            .vdel(
+                self.lease_name.clone(),
+                self.lease_holder_name.clone(),
+                None,
+                request_timeout,
+            )
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(state_store_error) => Err(state_store_error.into()),
+        }
```

Comment:
If the above comment about doing stuff on Ok doesn't make sense, then this can be simplified: 
```suggestion
         self
            .state_store
            .vdel(
                self.lease_name.clone(),
                self.lease_holder_name.clone(),
                None,
                request_timeout,
            )
            .await?
        Ok(())
```

Reply:
Done.